File: leetcode/00200_number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int R = grid.size();
        int C = grid[0].size();
        int count = 0;
        vector<pair<int, int>> offsets = vector<pair<int, int>>{
            {1, 0},
            {-1, 0},
            {0, 1},
            {0, -1},
        };
        
        queue<pair<int, int>> q;
        for (int r = 0; r < R; r++) {
            for (int c = 0; c < C; c++) {
                if (grid[r][c] == '0') {
                    continue;
                }
                count += 1;
                q.push(pair<int, int>{r, c});
                while (!q.empty()) {
                    pair<int, int> curr = q.front(); q.pop();
                    if (
                        !(0 <= curr.first && curr.first < R) ||
                        !(0 <= curr.second && curr.second < C) ||
                        grid[curr.first][curr.second] == '0'
                    ) {
                        continue;
                    }
                    grid[curr.first][curr.second] = '0';
                    for (auto offset : offsets) {
                        q.push({
                            curr.first + offset.first,
                            curr.second + offset.second,
                        });
                    }
                    
                }
            }
        }
        return count;
    }
};

File: leetcode/00200_number-of-islands/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        R, C = len(grid), len(grid[0])
        directions = [
            (1, 0),
            (-1, 0),
            (0, 1),
            (0, -1),
        ]
        count = 0

        def dfs(r, c):
            if not 0 <= r < R or not 0 <= c < C or grid[r][c] is not "1":
                return
            grid[r][c] = "0"
            for dr, dc in directions:
                dfs(r + dr, c + dc)

        for r in range(R):
            for c in range(C):
                if grid[r][c] == "0":
                    continue
                dfs(r, c)
                count += 1

        return count
